src/engine.rs

```diff
@@ -1,14 +1,100 @@
 use borsh::{BorshDeserialize, BorshSerialize};
 use evm::backend::{Apply, ApplyBackend, Backend, Basic, Log};
 use evm::executor::{MemoryStackState, StackExecutor, StackSubstateMetadata};
-use evm::{Config, CreateScheme, ExitError, ExitReason, ExitSucceed};
+use evm::ExitFatal;
+use evm::{Config, CreateScheme, ExitError, ExitReason};
 
 use crate::parameters::{FunctionCallArgs, NewCallArgs, ViewCallArgs};
 use crate::precompiles;
 use crate::prelude::{Address, Borrowed, Vec, H256, U256};
 use crate::sdk;
 use crate::storage::{address_to_key, storage_to_key, KeyPrefix};
-use crate::types::{bytes_to_hex, log_to_bytes, u256_to_arr, AccountId, NonceError};
+use crate::types::{bytes_to_hex, log_to_bytes, u256_to_arr, AccountId};
+
+/// Errors involving the nonce
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum NonceError {
+    /// Attempted to increment the nonce, but overflow occurred
+    NonceOverflow,
+    /// Account nonce did not match the transaction nonce
+    IncorrectNonce,
+}
+
+impl NonceError {
+    pub fn to_str(&self) -> &str {
+        use NonceError::*;
+        match self {
+            NonceOverflow => "ERR_NONCE_OVERFLOW",
+            IncorrectNonce => "ERR_INCORRECT_NONCE",
+        }
+    }
+}
+
+/// A result for nonces.
+pub type NonceResult<T> = Result<T, NonceError>;
+
+/// Errors with the EVM engine.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum EngineError {
+    /// Normal EVM errors.
+    EvmError(ExitError),
+    /// Fatal EVM errors.
+    EvmFatal(ExitFatal),
+    /// Revert of the transaction.
+    EvmRevert(Vec<u8>),
+}
+
+impl EngineError {
+    pub fn to_str(&self) -> &str {
+        use EngineError::*;
+        match self {
+            EvmError(ExitError::StackUnderflow) => "ERR_STACK_UNDERFLOW",
+            EvmError(ExitError::StackOverflow) => "ERR_STACK_OVERFLOW",
+            EvmError(ExitError::InvalidJump) => "ERR_INVALID_JUMP",
+            EvmError(ExitError::InvalidRange) => "ERR_INVALID_RANGE",
+            EvmError(ExitError::DesignatedInvalid) => "ERR_DESIGNATED_INVALID",
+            EvmError(ExitError::CallTooDeep) => "ERR_CALL_TOO_DEEP",
+            EvmError(ExitError::CreateCollision) => "ERR_CREATE_COLLISION",
+            EvmError(ExitError::CreateContractLimit) => "ERR_CREATE_CONTRACT_LIMIT",
+            EvmError(ExitError::OutOfOffset) => "ERR_OUT_OF_OFFSET",
+            EvmError(ExitError::OutOfGas) => "ERR_OUT_OF_GAS",
+            EvmError(ExitError::OutOfFund) => "ERR_OUT_OF_FUND",
+            EvmError(ExitError::Other(m)) => m,
+            EvmError(_) => "", // unused misc
```

Comment:
Is this reachable?

Reply:
Yes, there are some errors as part of `EvmError` that are flagged as unused. Not sure why they left them in as its totally dead code and nothing handles it.

Reply:
Can we put an `unreachable` guard here instead of an empty string?

Reply:
Sure.